### PR TITLE
Python 3.11 - fix majority of remaining test failures.

### PR DESCRIPTION
**Pull request by stuartarchibald**
This branch is based on the `py3.11` branch from 61e6a51.
* #8614 is included as patches c96fa09 and 34e2ed7 (the first in the series of patches in this PR).
* #8621 is included from 7c067fe through to 1af5b2c inclusive (with conflicts resolved).
* Additional patches are added based on feedback in the above PRs and also to match conventions adopted in the above PRs.

### CODE REVIEW

**Comment by stuartarchibald** [comment]
8543e6a recovers the testing fail/pass state as per https://github.com/numba/numba/pull/8621#issuecomment-1330484802.

**Comment by stuartarchibald** [comment]
Failing tests as of 8543e6a (potential reason/fix in brackets):
1. `numba.tests.test_exceptions.TestRaising.test_reraise_nopython` (think that this is because the Python 3.11 adjustments to exception handling don't know if there's a "live" handling region, error is `UnsupportedError(The re-raising of an exception is not yet supported`)).
2. `numba.tests.test_exceptions.TestRaising.test_reraise_objmode` (as above)
3. `numba.tests.test_profiler.TestProfiler.test_profiler` (Numba `C` code for tracing needs fixing).
4. `numba.tests.test_try_except.TestTryExceptOtherControlFlow.test_objmode` (unknown, error is: `AssertionError: 'unsupported control flow: with-context contains branches...'`).
5. ~`numba.tests.test_exceptions.TestRaising.test_user_code_error_traceback_nopython` (error message in 3.11 has changed to include markers, fix is trivial).~ fixed by c0cd45b.
6. `numba.tests.test_try_except.TestTryExceptOtherControlFlow.test_objmode_output_type` (error message changed, fix is potentially trivial).
7. ~`numba.tests.test_sort.TestTimsortArrays.test_merge_lo_hi` (this is potentially complicated, the bytecode offsets are being computed incorrectly in the case of an `EXTENDED_ARG`, further, Python 3.11 seems to use `EXTENDED_ARG` opcodes as jump targets.~ fixed by 1af5b2c.
8. ~`numba.tests.test_sort.TestTimsortArrays.test_run_timsort` (as above)~ fixed by 1af5b2c.
9. ~`numba.tests.test_sort.TestTimsortArrays.test_run_timsort_with_values` (as above)~ fixed by 1af5b2c.
10. `numba.tests.test_operators.TestMixedIntsOperatorModule.test_pow` (fails with, `AssertionError: when comparing -1261475310744950528 and -1261475310744950487`, there is something a little strange about the compilation of this function (uses `compile_isolated`)). Further debug notes... this looks like a float point issue, the test is right at the limit of `float64` precision.
11. `numba.tests.test_withlifting.TestLiftObj.test_case19_recursion` (unknown, error is `ValueError: The truth value of an array with more than one element is ambiguous.`, seems like something is compiling when it shouldn't be!).
12. `numba.tests.test_withlifting.TestLiftCall.test_liftcall5` (unknown, probably CF analysis related, error is `ValueError: Cannot add edge as dest node 34 not in nodes {0, 104}`)
13. `numba.tests.test_import.TestNumbaImport.test_no_accidental_warnings` (needs a `llvmlite` level fix, there's a `DeprecationWarning` coming from the FFI bindings' use of `importlib`.) xref: https://github.com/numba/llvmlite/pull/869#issuecomment-1327624103

File: numba/core/interpreter.py [hunk 1]
```diff
@@ -2859,23 +2859,36 @@ def op_JUMP_IF_TRUE(self, inst, pred):
...
-        predicate = self.store(value=isnone, name=pname)
-        branch = ir.Branch(cond=predicate, truebr=truebr, falsebr=falsebr,
+        name = f"$bool{inst.offset}"
+        gv_fn = ir.Global("bool", bool, loc=self.loc)
+        self.store(value=gv_fn, name=name)
+
+        callres = ir.Expr.call(self.get(name), (self.get(maybeNone),), (),
+                               loc=self.loc)
+
+        pname = f"$pred{inst.offset}"
```
Comment (stuartarchibald) [inline numba/core/interpreter.py:2887]: Note to self. Make these register names match those that are generated for < 3.11.

**Review by sklam — APPROVED** [review]
Yeah, This gets us down to 10 errors on Azure

File: numba/core/interpreter.py [hunk 2]
```diff
@@ -2859,23 +2859,36 @@ def op_JUMP_IF_TRUE(self, inst, pred):
...
-        falsebr = brs[not iftrue]
+        # branch pruning assumes true falls through and false is jump
+        truebr = inst.next
+        falsebr = inst.get_jump_target()
+
+        # this seems strange
+        if not iftrue:
+            op = BINOPS_TO_OPERATORS["is"]
+        else:
+            op = BINOPS_TO_OPERATORS["is not"]
```
Comment (sklam) [inline numba/core/interpreter.py:2870]: `inst.next` is the fallthrough branch. `inst.get_jump_target()` is the take-the-jump branch. need to fix up the `truebr` `falsebr` variable naming.

**Pull request merged**